notes/overriding.md

```diff
@@ -0,0 +1,676 @@
+   ```
+   However, dynamic scoping have a number of issues, and is usually considered
+   bad practice.
+ - **lexical scoping**: as currently, require self-referenced fields to be
+   defined locally. Note that thanks to contracts, one can require the existence
+   of a field without defining it. For example, we could write the previous
+   example as:
+   ```nickel
+   {a = b, b | Num} & {b = 1}
+   ```
+   This is also a better practice to explicitly state the fields whose presence
+   is required.
+
+### Priorities
+
+Merge is fundamentally commutative. The problem is, not all fields should be
+treated the same: some are default values that ought to be overridden, some are
+high-priority values that ought to override. The solution is to use a priority
+system, encoding the precedence of each value while retaining commutativity
+(only the priority annotation counts, not the order in which values are merged).
+
+Priorities have drawbacks. One is non-locality: the final result depends on
+other priorities of the field's values you are being merged with, potentially
+defined elsewhere. However, this can be mitigated by good defaults and an
+adapted set of priorities. For example:
+
+1. `default` is lowest than everything else (bottom)
+2. Integer priorities in the middle, with `0` being the default
+3. `force` is higher than everything else (top)
+
+Doing so, it's easy to just use `default` or `force` to override or be
+overridden by anything, without knowing the precise integer priority. Integer
+priorities still gives freedom with an infinite supply of levels if that is
+required, and the default priority is right in the middle.
+
+However, as noted in [#240](https://github.com/tweag/nickel/issues/240),
+configuration should be easily overridable, and the approach outlined here can
+become annoying because it may end up requiring either configurations to be
+written with `default` or `force` everywhere. A solution is to have additional
+recursive (or "leafy", or "push down") annotations, as described in
+[#279](https://github.com/tweag/nickel/issues/279):
+
+```nickel
+let someNeutraleConfig = {
+  foo = 1,
+  bar.baz = "stuff",
+  bar.blorg = false,
+}
+let defaulted | default rec = someNeutralConfig;
+
+// defaulted will evaluate to:
+// {
+//   foo | default = 1,
+//   bar = {
+//     baz | default = "stuff",
+//     bar.blorg | default = false,
+//   },
+// }
+// This is different from defaulted | default !
+
+// Same with force:
+let overrider | force rec = someNeutralConfig;
+```
+
+This way, an existing definition (arbitrarily complex: that could be the root of
+all Nixpkgs) can easily (and lazily) be turned into a full overriding or
+overridable configuration.
+
+### Custom merge functions
+
+Custom merge functions would be specified by a `merge` metavalue attribute. In
+order to enforce commutativity, they would receive their arguments in an
+indistinguishable order, such as `{lower: Dyn, higher: Dyn}`, just knowing which
```

Comment:
I assume that the custom merge functions will probably need to know when they are getting two terms of equal priority.